web/pandas/pdeps/0006-ban-upcasting.md

```diff
@@ -0,0 +1,149 @@
+# PDEP-6: Ban upcasting in setitem-like operations
+
+- Created: 23 December 2022
+- Status: Draft
+- Discussion: [#50402](https://github.com/pandas-dev/pandas/pull/50402)
+- Author: [Marco Gorelli](https://github.com/MarcoGorelli) ([original issue](https://github.com/pandas-dev/pandas/issues/39584) by [Joris Van den Bossche](https://github.com/jorisvandenbossche))
+- Revision: 1
+
+## Abstract
+
+The suggestion is that setitem-like operations would
+not change a ``Series``' dtype.
+
+Current behaviour:
+```python
+In [1]: ser = pd.Series([1, 2, 3], dtype='int64')
+
+In [2]: ser[2] = 'potage'
+
+In [3]: ser  # dtype changed to 'object'!
+Out[3]:
+0         1
+1         2
+2    potage
+dtype: object
+```
+
+Suggested behaviour:
+
+```python
+In [1]: ser = pd.Series([1, 2, 3])
+
+In [2]: ser[2] = 'potage'  # raises!
+---------------------------------------------------------------------------
+TypeError: Invalid value 'potage' for dtype int64
+```
+
+## Motivation and Scope
+
+Currently, pandas is extremely flexible in handling different dtypes.
+However, this can potentially hide bugs, break user expectations, and copy data
+in what looks like it should be an inplace operation.
+
+An example of it hiding a bug is:
+```python
+In [9]: ser = pd.Series(pd.date_range('2000', periods=3))
+
+In [10]: ser[2] = '2000-01-04'  # works, is converted to datetime64
+
+In [11]: ser[2] = '2000-01-04x'  # almost certainly a typo - but pandas doesn't error, it upcasts to object
+```
+
+The scope of this PDEP is limited to setitem-like operations which would operate inplace, such as:
```

Comment:
Not sure what you mean exactly, but a regular setitem is not inplace anymore. We never write into the underlying array since 1.3 I think.

Reply:
> We never write into the underlying array since 1.3 I think.

Worth clarifying this is for `DataFrame.__setitem__` only.  NOT for `Series.__setitem__` or any loc/iloc paths.

Reply:
Thx, should have been clearer.

Reply:
Right, so `DataFrame.loc[0, 'col1'] = 2.5` would be out-of-scope for this PDEP - thanks for your comments, I've added that to the examples to clarify

Reply:
I think this could be confusing, because as I understand it, the proposal would have the following behavior:
```python
s = pd.Series([1,2,3])   # s will have dtype int
df = pd.DataFrame({"x": s})   # column "x" will have dtype int
s.iloc[1] = 2.5  # This would raise
df.loc[1, "x"] = 2.5 #  This would upcast the column "x" to have dtype float
```
If this is what is proposed, this inconsistent behavior will be VERY confusing.

Reply:
That's correct

I've retitled the issues as `ban upcasting in Series setitem-like operations`, to make it clear that this only for Series.

"setitem-like operations won't change a Series' dtype" is simple, and doesn't seem confusing - `df.loc[1, 'x']` isn't an operation on a Series afterall

There's a risk here that, if we can't do everything, then we end up doing nothing, and I'm hoping we can avoid that

Reply:
id suggest the scope be "potentially inplace operations" which would include `df.loc[1, "x"] = 2.5`

Reply:
> Right, so `DataFrame.loc[0, 'col1'] = 2.5` would be out-of-scope for this PDEP

I also don't fully understand what the rationale would be of leaving out this case. This currently updates the column inplace (if it exists / if the dtype matches), so IMO would fall under the setitem-like operations

Reply:
OK, I have to read better first ;) I see your last comment gave this clarification as rationale:

> to make it clear that this only for Series.

But personally, I am not sure the distinction between Series and DataFrame column is that useful. You could also phrase it like "setitem-like operations won't change a Series' or DataFrame column's dtype", which is still simple and IMO more consistent and even easier to reason about (why would updating a Series inplace behave different as updating a DataFrame column inplace? Logically speaking both operations are equivalent to me as a user)

Reply:
Agreed, thanks for your comments - I've included `DataFrame.loc[0, 'col1'] = 2.5` in the scope